docs/source/architecture/events_and_transactions_receipts.rst

```diff
@@ -0,0 +1,343 @@
+   :align: center
+   :alt: Event Subsystem
+
+Events
+======
+
+Events will be represented with the following protobuf message:
+
+.. code-block:: protobuf
+
+  message Event {
+    // Used to subscribe to events and servers as a hint for how to deserialize
+    // event_data and what pairs to expect in attributes.
+    string event_type = 1;
+
+    // Transparent data defined by the event_type.
+    message Attribute {
+      string key = 1;
+      string value = 2;
+    }
+    repeated Attribute attributes = 2;
+
+    // Opaque data defined by the event_type.
+    bytes  data = 3;
+  }
+
+
+Events are extracted from other data structures such as blocks or
+transaction receipts. In order to treat this extraction uniformly, an
+EventExtractor interface is used. This interface is implemented for
+each event source. The EventExtractor interface takes a list of
+EventSubscriptions and will only generate events that are in the union of all
+subscriptions. An event is "in a subscription" if the event's event_type field
+matches the subscription's event_type field and, if the subscription has any
+filters, the key-value pair of the filter matches the key-value pair of one of
+the key-value pairs in the event's meta_data field. EventExtractor
+implementations should be lightweight and reusable across the system.
+
+.. code-block:: python
+
+  interface EventExtractor:
+    // Construct all the events of interest by taking the union of all subscriptions.
+    // One extractor should be created for each input source that events can be
+    // extracted from. This input source should be passed to the implementation through
+    // the constructor.
+    extract(list<EventSubscription> subscriptions) -> list<Event>
+
+  // If no subscriptions of a given event_type are passed to EventExtractor.extract,
+  // the extractor does not need to return events of that type.
+  class EventSubscription:
+    string event_type
+    list<EventFilter> filters
+
+  class EventFilter(object, metaclass=ABCMeta):
+      """Represents a subset of events within an event type."""
+
+      def __init__(self, key, match_string):
+          self.match_string = match_string
+          self.key = key
+
+      def __eq__(self, other):
+          if not isinstance(other, self.__class__):
+              return False
+          return self.key == other.key \
+              and self.match_string == other.match_string
+
+      def __contains__(self, event):
+          return self.matches(event)
+
+      @abstractmethod
+      def matches(self, event):
+          """Returns whether the event passes this filter."""
+          raise NotImplementedError()
+
+An EventBroadcaster manages external event subscriptions and forwards events to
+subscribers as they occur. In order for the EventBroadcaster to learn about
+events, the ChainController class implements the Observer pattern. The
+ChainController acts as the subject and observers of the ChainController is
+a ChainObserver interface.
```

Comment:
Change "is a ChainObserver interface" to "implement the ChainObserver interface"